pkg/coscheduling/coscheduling.go

```diff
@@ -0,0 +1,268 @@
+
+const (
+	// Name is the name of the plugin used in Registry and configurations.
+	Name                 = "coscheduling"
+	PodGroupName         = "pod-group.scheduling.sigs.k8s.io/name"
+	PodGroupMinAvailable = "pod-group.scheduling.sigs.k8s.io/min-available"
+	PermitWaitingTime    = 1 * time.Second
+)
+
+// Name returns name of the plugin. It is used in logs, etc.
+func (cs *Coscheduling) Name() string {
+	return Name
+}
+
+// New initializes a new plugin and returns it.
+func New(_ *runtime.Unknown, handle framework.FrameworkHandle) (framework.Plugin, error) {
+	podLister := handle.SharedInformerFactory().Core().V1().Pods().Lister()
+	//podGroupInfos := make(map[string]PodGroupInfo)
+	return &Coscheduling{FrameworkHandle: handle,
+		PodLister: podLister,
+	}, nil
+}
+
+// Less are used to sort pods in the scheduling queue.
+// 1. compare the priority of pod
+// 2. compare the timestamp of the initialization time of PodGroup
+// 3. compare the SeqNum
+func (cs *Coscheduling) Less(podInfo1 *framework.PodInfo, podInfo2 *framework.PodInfo) bool {
+	pod1 := podInfo1.Pod
+	pod2 := podInfo2.Pod
+	priority1 := pod.GetPodPriority(pod1)
+	priority2 := pod.GetPodPriority(pod2)
+
+	if priority1 != priority2 {
+		return priority1 > priority2
+	}
+
+	pgInfo1 := cs.getPodGroupInfo(podInfo1)
+	pgInfo2 := cs.getPodGroupInfo(podInfo2)
+	time1 := pgInfo1.Timestamp
+	time2 := pgInfo2.Timestamp
+
+	if pgInfo1.SeqNum > 0 && pgInfo2.SeqNum > 0 && time1.Equal(time2) {
+		return pgInfo1.SeqNum < pgInfo2.SeqNum
+	}
+	return time1.Before(time2)
+}
+
+func (cs *Coscheduling) getPodGroupInfo(p *framework.PodInfo) *PodGroupInfo {
+	podGroupName, min := GetPodGroupLabels(p.Pod)
+	if podGroupName != "" {
+		key := fmt.Sprintf("%v/%v", p.Pod.Namespace, podGroupName)
+		pgInfo, ok := cs.PodGroupInfos.Load(key)
+		if !ok {
+			pgInfo = &PodGroupInfo{
+				Name:         podGroupName,
+				Timestamp:    p.InitialAttemptTimestamp,
+				SeqNum:       cs.GetSeqNum(),
+				MinAvailable: min,
+			}
+			cs.PodGroupInfos.Store(key, pgInfo)
+		}
+		return pgInfo.(*PodGroupInfo)
+	}
+
+	// If the pod is regular pod, return object of PodGroupInfo but not store in PodGroupInfos.
+	// The purpose is to facilitate unified comparison.
+	return &PodGroupInfo{Timestamp: p.InitialAttemptTimestamp, SeqNum: 0}
+}
+
+// PreFilter validates that if the total number of pods belonging to the same `PodGroup` is less than `minAvailable`.
+// If so, the scheduling process will be interrupted directly to avoid the partial Pods holding system resources
+// until a timeout. It will reduce the overall scheduling time for the whole group
+func (cs *Coscheduling) PreFilter(ctx context.Context, state *framework.CycleState, p *v1.Pod) *framework.Status {
+	podGroupName, minAvailable := GetPodGroupLabels(p)
+	if podGroupName == "" || minAvailable <= 1 {
+		return framework.NewStatus(framework.Success, "")
+	}
+
+	total := cs.calculateTotalPods(podGroupName, p.Namespace)
+	if total < minAvailable {
```

Comment:
The first several pods of a group may be Unschedulable in PreFilter because when they entering PreFilter, some other pods of the group may havn't been created , then all pods of the group will be Unschedulable in Permit.It's a long time until they get the next chance to be scheduled.The way of counting total pods of a group needs improvement, otherwise I prefer disable PreFilter.

Reply:
I'm afraid this is one of the limitations of current scheduler framework - which doesn't support multi-queue, and also doesn't expose that as an extension point yet. So here I read it as a hack to work around that. Otherwise, we may have to start the 1st PG Pod's scheduling cycle right away and that may leads to inefficient waiting.

Reply:
It looks like people should be able to disable this PreFilter if they find it detrimental rather than useful. Am I correct?
If so, we should add this to the README.

Reply:
this was not handled